### PR TITLE
Add combinators to STM

### PR DESCRIPTION
**Pull request by sken77**
andThen >>>
compose <<<
eventually
catchAll
flatMapError
flattenErrorOption

tests are next

### CODE REVIEW

File: core/shared/src/main/scala/zio/stm/ZSTM.scala [hunk 1]
```diff
@@ -219,12 +252,25 @@ final class ZSTM[-R, +E, +A] private[stm] (
       case TExit.Retry      => ZSTM.retry
     }
 
+  /**
+   * Creates a composite effect that represents this effect followed by another
+   * one that may depend on the error produced by this one.
+   */
+  final def flatMapError[R1 <: R, E2](f: E => ZSTM[R1, Nothing, E2])(implicit ev: CanFail[E]): ZSTM[R1, E2, A] =
+    flipWith(_ flatMap f)
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/stm/ZSTM.scala]: Probably more efficient in terms of `foldM`.

**Comment by sken77** [comment]
test for `eventually` is timing out and im not sure why

**Comment by adamgfraser** [comment]
I think it because `orElse` resets the journal before continuing with the right hand side effect, which is somewhat surprising to me.

@mijicd Is this intended behavior?

**Comment by mijicd** [comment]
Exactly, in case of failure/retry of left, journal will be nuked. That's the intended behavior.

**Comment by adamgfraser** [comment]
Hmmm. I wonder if we should think about that or at least document it more because it is significantly different than in ZIO. In ZIO you could define `orElse(zio)` as `foldM(_ => zio, ZIO.succeed)` whereas the semantic here is quite different. STM does provide different functionality so don't want to shoehorn it into ZIO but it does seem like it could be surprising.

**Comment by sken77** [comment]
It works fine if i use `foldM` instead of `orElse`

**Comment by adamgfraser** [comment]
Yes, that's definitely a good solution for this PR.

**Comment by sken77** [comment]
I'm not sure how would i go about testing `andThen` and `compose` couldnt find an example on `ZIOSpec`

**Comment by adamgfraser** [comment]
Try creating two STM effects that access say an `Int` in the environment and do something with it, then glue them together using `andThen` and `compose`, and test that if you provide them with a value for the environment you get the expected value back.

**Comment by sken77** [comment]
https://github.com/zio/zio/issues/2802

**Pull request merged**